custom_components/adaptive_lighting/switch.py

```diff
@@ -238,57 +237,102 @@ def _split_service_data(service_data, adapt_brightness, adapt_color):
+) -> list[AdaptiveSwitch]:
     _LOGGER.debug(
-        "Called 'adaptive_lighting.set_manual_control' service with '%s'",
+        "Called '_parse_service_args' with service data:\n'%s'",
         service_call.data,
     )
-    if service_call.data[CONF_MANUAL_CONTROL]:
-        for light in all_lights:
-            switch.turn_on_off_listener.manual_control[light] = True
-            _fire_manual_control_event(switch, light, service_call.context)
-    else:
-        switch.turn_on_off_listener.reset(*all_lights)
-        # pylint: disable=protected-access
-        if switch.is_on:
-            await switch._update_attrs_and_maybe_adapt_lights(
-                all_lights,
-                transition=switch._initial_transition,
-                force=True,
-                context=switch.create_context("service", parent=service_call.context),
+    data = service_call.data
+    lights = data[CONF_LIGHTS]
+    switch_unique_id: list[str] | None = data.get("entity_id")
+    if not lights and not switch_unique_id:
+        # alternatively just loop and call the service for every switch.
+        _LOGGER.debug(
+            "adaptive-lighting: If you were trying to adapt every single light on every single"
+            " switch, please inform the developers at"
+            " https://github.com/basnijholt/adaptive-lighting of your use case."
+            " Currently you must pass either the adapt main switch, or the lights to every"
+            " service call. adaptive-lighting will now raise a soft error for you in the meantime."
+        )
+        _LOGGER.error(
+            "bad service data to adaptive-lighting service call -"
+            " you must pass either an entity_id or a light. Service data:\n%s",
+            service_call.data,
+        )
+        raise ValueError(
+            "adaptive-lighting: No switch or light was passed to service call."
+        )
+
+    if switch_unique_id is not None:
+        if len(switch_unique_id) > 1 and lights:
+            _LOGGER.error(
+                "bad service data: cannot pass multiple switch entities while also passing"
+                " any lights. Got %s",
+                service_call.data,
             )
+            raise ValueError(
+                "adaptive-lighting: Multiple switches passed with lights argument"
+            )
+        switches = []
+        for unique_id in switch_unique_id:
+            ent_reg = entity_registry.async_get(hass)
+            ent_entry = ent_reg.async_get(unique_id)
+            config_id = ent_entry.config_entry_id
+            _LOGGER.debug(
+                "_parse_service_args: switches: %s, ent_reg: %s, ent_entry: %s, config_id: %s",
+                switches,
+                ent_reg,
+                ent_entry,
+                config_id,
+            )
+            switches.append(hass.data[DOMAIN][config_id]["instance"])
+        return switches
+    elif lights:
+        switch = _find_switch_with_lights(hass, lights, service_call)
```

Comment:
I have factored the code in the `_find_switch_with_lights` function. However, I believe there exists a bug in code. Previously `switch` could be `None`, `list[AdaptiveSwitch]` and `AdaptiveSwitch`. I believe elsewhere it always assumes a list of switches is returned.